# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: widget/ContentCache.cpp
# Commit: b4c05874394f
# Full Hash: b4c05874394f31b10e8baca8a7a36c34809c2b82
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-02-17 21:21:27
# Description:
#   Bug 1755438 - Make `ContentCacheInParent::GetTextRect` check `mSelection` before referring it r=m_kato
#   
#   Unfortunately, the adding test cannot reproduce the crash, but I have no idea
#   how to make `mSelection` being `Nothing` until next native event loop.  Anyway,
#   it's worthwhile to add new tests.
# ==============================================================================

diff -r 554b03ae80a9 -r b4c05874394f widget/ContentCache.cpp
--- a/widget/ContentCache.cpp	Thu Feb 17 06:33:39 2022 +0000
+++ b/widget/ContentCache.cpp	Thu Feb 17 08:53:35 2022 +0000
@@ -880,6 +880,8 @@
               ("0x%p HandleQueryContentEvent(aEvent={ "
                "mMessage=eQueryEditorRect }, aWidget=0x%p)",
                this, aWidget));
+      // XXX This query should fail if no editable elmenet has focus.  Or,
+      //     perhaps, should return rect of the window instead.
       aEvent.EmplaceReply();
       aEvent.mReply->mFocusedWidget = aWidget;
       aEvent.mReply->mRect = mEditorRect;
@@ -968,8 +970,13 @@
 
   if (mTextRectArray.isNothing() || !mTextRectArray->HasRects()) {
     // If there are no rects in mTextRectArray, we should refer the start of
-    // the selection because IME must query a char rect around it if there is
-    // no composition.
+    // the selection if there is because IME must query a char rect around it if
+    // there is no composition.
+    if (mSelection.isNothing()) {
+      // Unfortunately, there is no data about text rect...
+      aTextRect.SetEmpty();
+      return false;
+    }
     aTextRect = mSelection->StartCharRect();
     return !aTextRect.IsEmpty();
   }